access/email/app.go

```diff
@@ -0,0 +1,483 @@
+		}
+		pluginData.EmailThreads = threadsSent
+		return pluginData, true
+	})
+	return trace.Wrap(err)
+}
+
+// sendReviews sends notifications on a request updates (new accept/decline review, review expired)
+func (a *App) sendReviews(ctx context.Context, reqID string, reqData RequestData, reqReviews []types.AccessReview) error {
+	var oldCount int
+	var threads []EmailThread
+
+	ok, err := a.modifyPluginData(ctx, reqID, func(existing *PluginData) (PluginData, bool) {
+		if existing == nil {
+			return PluginData{}, false
+		}
+
+		if threads = existing.EmailThreads; len(threads) == 0 {
+			return PluginData{}, false
+		}
+
+		count := len(reqReviews)
+		if oldCount = existing.ReviewsCount; oldCount >= count {
+			return PluginData{}, false
+		}
+		pluginData := *existing
+		pluginData.ReviewsCount = count
+		return pluginData, true
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if !ok {
+		logger.Get(ctx).Debug("Failed to post reply: plugin data is missing")
+		return nil
+	}
+	reviews := reqReviews[oldCount:]
+	if len(reviews) == 0 {
+		return nil
+	}
+
+	errors := make([]error, 0, len(reviews))
+	for _, review := range reviews {
+		threadsSent, err := a.bot.SendReview(ctx, threads, reqID, reqData, review)
+		if err != nil {
+			errors = append(errors, err)
+		}
+		logger.Get(ctx).Infof("New review for request %v by %v is %v", reqID, review.Author, review.ProposedState.String())
+		logSentThreads(ctx, threadsSent)
+	}
+
+	return trace.NewAggregate(errors...)
+}
+
+// sendStatusUpdates updates the messages status and sends message when request has been resolved
+func (a *App) sendStatusUpdates(ctx context.Context, reqID string, resolution Resolution) error {
+	log := logger.Get(ctx)
+
+	var pluginData PluginData
+	ok, err := a.modifyPluginData(ctx, reqID, func(existing *PluginData) (PluginData, bool) {
+		// If plugin data is empty or missing slack message timestamps, we cannot do anything.
+		if existing == nil {
+			return PluginData{}, false
+		}
+		if pluginData = *existing; len(pluginData.EmailThreads) == 0 {
+			return PluginData{}, false
+		}
+
+		// If resolution field is not empty then we already resolved the incident before. In this case we just quit.
+		if pluginData.RequestData.Resolution.Tag != Unresolved {
+			return PluginData{}, false
+		}
+
+		// Mark plugin data as resolved.
+		pluginData.Resolution = resolution
+		return pluginData, true
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if !ok {
+		log.Debug("Failed to update messages: plugin data is missing")
+		return nil
+	}
+
+	reqData, threads := pluginData.RequestData, pluginData.EmailThreads
+
+	threadsSent, err := a.bot.SendResolveStatus(ctx, threads, reqID, reqData)
+	logSentThreads(ctx, threadsSent)
+
+	log.Infof("Marked request as %s and sent emails!", resolution.Tag)
+
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// modifyPluginData performs a compare-and-swap update of access request's plugin data.
```

Comment:
From #295 

```suggestion
// modifyPluginData performs a compare-and-swap update of access request's plugin data.
// Callback function parameter is nil if plugin data hasn't been created yet.
// Otherwise, callback function parameter is a pointer to current plugin data contents.
// Callback function return value is an updated plugin data contents plus the boolean flag
// indicating whether it should be written or not.
// Note that callback function fn might be called more than once due to retry mechanism baked in
// so make sure that the function is "pure" i.e. it doesn't interact with the outside world:
// it doesn't perform any sort of I/O operations so even things like Go channels must be avoided.
// Indeed, this limitation is not that ultimate at least if you know what you're doing.
```

Reply:
Done